tools/build_defs/framework.bzl

```diff
@@ -0,0 +1,284 @@
+#                            include_dir = depset([out_files.out_include_dir]),
+                            out_binary_files = depset(out_files.out_binary_files),
+                            pkg_config_dir = out_files.out_pkg_dir or []),
+            cc_common.create_cc_skylark_info(ctx=ctx),
+            cc_info.compilation_info,
+            cc_info.linking_info,
+]
+
+def _value(value, default_value):
+  if (value):
+    return value
+  return default_value
+
+def _depset(item):
+  if item == None:
+    return depset()
+  return depset([item])
+
+def _list(item):
+  if item:
+    return [item]
+  return []
+
+def _copy_deps_and_tools(files):
+  groups = {
+      "lib": files.libs,
+      "include": files.headers,
+      "lib/pkgconfig": files.pkg_configs,
+      "bin": files.tools_files,
+  }
+  list = []
+  for key in groups.keys():
+    group_files = groups[key]
+    if len(group_files) == 0:
+      continue
+
+    list += ["mkdir -p $EXT_BUILD_DEPS/" + key]
+    paths_to_copy = []
+    for file in group_files:
+      if (type(file) == "string"):
+        paths_to_copy += [file]
+      else:
+        paths_to_copy += [file.path]
+
+    # include directories would be mentioned both in headers and system_includes sections, avoid duplication
+    paths_to_copy = collections.uniq(paths_to_copy)
+
+    for path in paths_to_copy:
+      list += ["copy_to_dir $EXT_BUILD_ROOT/{} $EXT_BUILD_DEPS/{}".format(path, key)]
+
+  list += ["define_absolute_paths $EXT_BUILD_ROOT/bin $EXT_BUILD_ROOT/bin"]
+  list += ["path $EXT_BUILD_ROOT/bin"]
+  list += ["export PKG_CONFIG_PATH=$EXT_BUILD_ROOT/lib/pkgconfig"]
+
+  return list
+
+def _check_file_name(var, name):
+  if (len(var) == 0):
+    fail("{} can not be empty string.".format(name.capitalize()))
+
+  if (not var[0:1].isalpha()):
+    fail("{} should start with a letter.".format(name.capitalize()))
+  if (not var.isalnum()):
+    fail("{} should be alphanumeric.".format(name.capitalize()))
+
+def _define_outputs(ctx, lib_name):
+  if (ctx.attr.static_library == None and ctx.attr.dynamic_library == None and ctx.attr.interface_library == None and len(ctx.attr.binaries_names) == 0):
+    fail("One of \"out_lib_name\" or \"out_bin_name\" attributes must be provided.")
+
+  _check_file_name(lib_name, "Library name")
+
+  out_binary_files = []
+  for file in ctx.attr.binaries_names:
+    out_binary_files += [_declare_out(ctx, lib_name, ctx.attr.out_bin_dir, file)]
+
+  out_pkg_dir = None
+  if ctx.attr.out_pkg_config_dir:
+    out_pkg_dir = ctx.actions.declare_file("/".join([lib_name, ctx.attr.out_pkg_config_dir]))
+
+  wrapper = ctx.actions.declare_directory(lib_name)
+  out_include_dir = ctx.actions.declare_directory(lib_name + "/" + ctx.attr.out_include_dir)
+  out_bin_dir = ctx.actions.declare_directory(lib_name + "/" + ctx.attr.out_bin_dir)
+  out_lib_dir = ctx.actions.declare_directory(lib_name + "/" + ctx.attr.out_lib_dir)
+
+  libraries = struct(
```

Comment:
Consider declaring a provider for this (even if it is not returned from a rule). That will help document what goes into this struct